packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -145,63 +139,71 @@ export default createRule<Options, MessageIds>({
         [],
       );
     }
-    function reportDuplicate(
-      duplicateConstituent: {
-        duplicated: TSESTree.TypeNode;
-        duplicatePrevious: TSESTree.TypeNode;
-      },
-      parentNode: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    function report(
+      messageId: MessageIds,
+      node: TSESTree.TypeNode,
+      data?: Record<string, unknown>,
     ): void {
-      const beforeTokens = context.sourceCode.getTokensBefore(
-        duplicateConstituent.duplicated,
-        { filter: token => token.value === '|' || token.value === '&' },
-      );
-      const beforeUnionOrIntersectionToken =
-        beforeTokens[beforeTokens.length - 1];
-      const bracketBeforeTokens = context.sourceCode.getTokensBetween(
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const beforeUnionOrIntersectionToken = sourceCode
+        .getTokensBefore(node, {
+          filter: token => ['|', '&'].includes(token.value),
+        })
+        .at(-1)!;
+      const bracketBeforeTokens = sourceCode.getTokensBetween(
```

Comment:
[Bug] Rule throws error on this code:

```
getTokensBetween@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:451479
o@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:1770041
TSUnionType@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:1770599
ad</Y/</</<@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:1394247
Fp</t.exports/emit/<@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:1350213
emit@https://deploy-preview-9479--typescript-eslint.netlify.app/sandbox/index.js:1:1350200...
```

```ts
function a(arg?: undefined | string) {}
```

[playground](https://deploy-preview-9479--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.ts&code=GYVwdgxgLglg9mABAQwBTIE4HMD8AuRcAEwFNgYwSjEAfRAZygwqwEpEBvAXyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false)

Reply:
Good catch, it is now possible to report on the first element of a union or intersection, where that wasn't possible before